### PR TITLE
[WIP] rust: use workspace dependencies

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -80,6 +80,267 @@ resolver = "2"
...
+typemap_rev = "0.3.0"
+uncased = "0.9.7"
+unicode-width = "0.1.10"
+url = "2.3.1"
+uuid = "1.2.2"
+walkdir = "2.3.2"
+yansi = "0.5.1"
+zeroize = "1.5.7"
+
+mz-adapter = { path = "src/adapter" }
```
Comment (danhhz) [inline Cargo.toml:277]: do we get any particular benefit from putting all the path deps here, too? if not, i'm inclined to say we should leave them inlined

Comment (ParkMyCar) [reply Cargo.toml:277]: I only separated them because that's what we used to do with imports, I'll inline them!

File: Cargo.toml [hunk 2]
```diff
@@ -80,6 +80,266 @@ resolver = "2"
...
+syn = "1.0.107"
+sysctl = "0.5.4"
+sysinfo = "0.27.2"
+tar = "0.4.38"
+tempfile = "3.3.0"
+termcolor = "1.1.3"
+thiserror = "1.0.37"
+tiberius = { version = "0.11.3", default-features = false }
+tikv-jemalloc-ctl = "0.5.0"
+tikv-jemallocator = "0.5.0"
```
Comment (antiguru) [inline Cargo.toml:314]: These are optional dependencies. It would be great if we could keep it this way because several memory profiling tools have a hard time with jemalloc and work much better with the system allocator.

Comment (antiguru) [reply Cargo.toml:314]: (This is also why the lint fails)

File: src/expr/Cargo.toml [hunk 3]
```diff
@@ -11,56 +11,56 @@ name = "like_pattern"
...
+serde = { workspace = true, features = ["derive"] }
+serde_json = { workspace = true }
+serde_regex = { workspace = true }
+sha1 = { workspace = true }
+sha2 = { workspace = true }
+tracing = { workspace = true }
+uncased = { workspace = true }
+uuid = { workspace = true, features = ["v5"]}
+proptest = { workspace = true, default-features = false, features = ["std"] }
+proptest-derive = { workspace = true, features = ["boxed_union"]}
```
Comment (aalexandrov) [inline src/expr/Cargo.toml:51]: If we can define features in addition to the version number in the workspace `Cargo.toml`, I think it makes sense to also mandate `features = ["boxed_union"]` by default (see #16809).

Comment (benesch) [reply src/expr/Cargo.toml:51]: Yep, looks like you can specify features on the workspace dependency that get unioned with each package's features.

File: Cargo.toml (hunk 2 above)
Comment (benesch) [reply Cargo.toml:314]: Huh, I'm surprised this doesn't work as is. AFAICT the dependencies are still optional in the package itself; it's just that if they _are_ enabled they're meant to use the version here.

**Comment by ParkMyCar** [comment]
Moving this to the draft stage because I realized workspace dependencies don't work well with default features, I filed an issue against Cargo here https://github.com/rust-lang/cargo/issues/12162.

tl;dr if you inherit a dependency from the workspace, there is no way to disable default features for that dependency, without also disabling them for the entire workspace. What you need to do is first disable them for the entire workspace and then enable them in the crates which you want them, which is presumably most.

File: Cargo.toml (hunk 2 above)
Comment (ParkMyCar) [reply Cargo.toml:314]: Totally agree that we should keep these as optional deps, it was through investigating this lint that I found default-features don't really work with workspace dependencies. I filed https://github.com/rust-lang/cargo/issues/12162 to discuss this

Comment (benesch) [reply Cargo.toml:314]: The described behavior seems workable for us. Seems like it'll be annoying for the author of this PR (:D), but not too annoying to deal with once it's in place. There are very few crates where we want `default-features = false` in some places but not others; usually every Materialize crates wants to turn of the default features or none.
